scripts/generate-import-linter-config.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/env python3
+import argparse
+import configparser
+import importlib
+import pkgutil
+from collections.abc import ItemsView
+from typing import Literal, cast
+
+from typing_extensions import NotRequired, TypedDict
+
+
+AlertingLevel = Literal["error", "warn", "none"]
+ValueType = AlertingLevel | str | list[str]
+
+
+class Contract(TypedDict):
+    name: str
+    type: Literal["layers", "independence", "forbidden"]
+    layers: NotRequired[list[str]]
+    modules: NotRequired[list[str]]
+    ignore_imports: NotRequired[list[str]]
+    unmatched_ignore_imports_alerting: NotRequired[AlertingLevel]
+    source_modules: NotRequired[list[str]]
+    forbidden_modules: NotRequired[list[str]]
+
+
+NOTICE = (
+    "# This file has been generated by ./scripts/generate-import-linter-config.py,"
+    " do not edit by hand!"
+)
+CONTRACTS = "importlinter:contracts"
+STATIC_CONFIG = {
+    "importlinter": {
+        "root_package": "ggshield",
+        "include_external_packages": True,
+    },
+    CONTRACTS: [
+        {
+            "name": "ggshield-layers",
+            "type": "layers",
+            "layers": [
+                "ggshield.cmd|%",
+                "ggshield.verticals|%",
+                "ggshield.core",
+                "ggshield.utils | pygitguardian",
+            ],
+            "ignore_imports": [
+                "ggshield.cmd.main -> ggshield.cmd.**",
+                "ggshield.cmd.** -> ggshield.cmd.utils.*",
+            ],
+            "unmatched_ignore_imports_alerting": "warn",
+        },
+        {
+            "name": "verticals-cmd-transversals",
+            "type": "forbidden",
+            "source_modules": [
+                "ggshield.cmd.%",
+            ],
+            "forbidden_modules": [
+                "ggshield.verticals.%",
+            ],
+            "ignore_imports": [
+                "ggshield.cmd.main -> ggshield.verticals.**",
```

Comment:
I think `main` should not be allowed to import from `ggshield.verticals`. At least it does not need to right now.

This also made me realize we should move `main` out of `cmd`, because right now the script generates rules to allow `cmd.main` to import from `verticals.main`, which does not (and should not) exist. But we can do so in a separate PR.

Similarly, the script generates ignore rules to allow `cmd.utils` to import from `verticals.utils` which does not and should not exist either.

Reply:
By default, import rules are transitive, If `main` imports `cmd.hmsl` and `cmd.hmsl` imports `verticals.hmsl` then `main` import `verticals.hmsl`. We can be changed but only for the _forbidden_ contract, the _layers_ contract always allows indirect imports.

The script generates the config according to a simple set of rules, we can complexify it but is this worth the price?